librz/arch/isa/msp430/msp430_instruction_parse.inc

```diff
@@ -0,0 +1,195 @@
+#include <rz_types.h>
+#include "msp430_disas.h"
+#include "msp430_il.h"
+
+static ut8 parse_word_sizedness(ut8 opcode, char *instr_str);
+static Msp430SourceAddressingMode parse_src(char *src_str, RZ_OUT ut32 *src, RZ_OUT int *dst_offset);
+static Msp430DestinationAddressingMode parse_dst(char *dst_str, RZ_OUT ut32 *dst);
+
+void rz_msp430_instruction_from_command(RZ_OUT RZ_NONNULL Msp430Instruction *op, RZ_NONNULL struct msp430_cmd *cmd) {
+	op->itype = cmd->type;
+	op->iopcode = cmd->opcode;
+	op->word_sized = parse_word_sizedness(op->iopcode, cmd->instr);
+
+	if (op->itype == MSP430_TWOOP) {
+		int dst_offset;
+		op->src_mode = parse_src(cmd->operands, &(op->src), &dst_offset);
+		op->dst_mode = parse_dst(cmd->operands + dst_offset, &(op->dst));
+	} else if (op->itype == MSP430_ONEOP) {
+		op->dst_mode = parse_dst(cmd->operands, &(op->dst));
+	}
+}
+
+static ut8 parse_word_sizedness(cut8 opcode, char *instr_str) {
+	// the sign extend is special: it always reads a byte but writes a word
+	// this is the only reason there are 2 flags for the sizedness of a read and the sizedness of a write
+	if (opcode == MSP430_SXT) {
+		return WORD_SIZED_WRITE;
+	}
+
+	char *curr = instr_str;
+	while (*curr != '\0' && *curr != '.') {
+		curr++;
+	}
+
+	if (*curr == '\0') {
+		return WORD_SIZED_READ | WORD_SIZED_WRITE;
+	}
+
+	rz_warn_if_fail(*(curr + 1) == 'b');
+	return 0; // byte sized read and write
+}
+
+static ut16 parse_4digit_hex(char *num);
+static ut16 parse_immediate(char *imm);
+static ut16 parse_reg(char *reg);
+static ut16 parse_hex(char *num, int *end_offset);
+
+static Msp430SourceAddressingMode parse_src(char *src_str, RZ_OUT ut32 *src, RZ_OUT int *dst_offset) {
```

Comment:
Instead of parsing string, I wonder if it's better to modify the `librz/arch/isa/msp430/msg430_disas.c` instead and provide this kind of information as a C structure, similarly how other disassemblers do it, e.g. capstone-based ones.

Reply:
Parsing the operands is definitely pretty ugly, but it's way easier to do than messing with the internals of the disassembler.

Someone, very likely me, will probably touch the MSP430 plugin soon to add assembly anyway (there is no MSP430 assembler at the moment, I'm using an external Python script I found somewhere on Github to obtain the binary of instructions), that will be a suitable moment to rethink `disas.c`..

Reply:
I recommend focusing on improving the disassembly, analysis, and RzIL first and foremost before spending time adding the assembler. These are higher priority, and improving disassembler to provide instruction details for the analysis is much more important than writing an assembler.

Reply:
You're actually right, the parsing code is a constant source of overflow headache and bugs, it keeps making AddressSanitizer complain about illegal reads. I decided to remove the problem from its root and delete the code, I will instead make the disassembler bubble up the information I need.